File: features/share/graphql/golang1.x/mutation.go
package main

import (
	"context"
	"flag"
	"fmt"
	"os"
	"time"

	"github.com/machinebox/graphql"
)

func main() {
	// Shell variables for the Warehouse node
	api := os.Getenv("API")
	api_key := os.Getenv("API_KEY")

	itemNamePtr := flag.String("itemname", "Thing 1", "inventory name")
	quantityPtr := flag.Int("quantity", 1, "item quantity")

	flag.Parse()

	// Create a client (safe to share across requests)
	client := graphql.NewClient(api)

	getId := graphql.NewRequest(`
	  query q($itemName:String) {
		list_InventoryItems(filter: {itemName: {eq: $itemName}}) {
		  _InventoryItems {
			_id
		  }
		}
	  }
	`)

	getId.Var("itemName", *itemNamePtr)
	getId.Header.Set("x-api-key", api_key)

	var getIdResponse struct {
		List_InventoryItems struct {
			InventoryItems []struct {
				Id string `json:"_id"`
			} `json:"_InventoryItems"`
		} `json:"list_InventoryItems"`
	}

	if err := client.Run(context.Background(), getId, &getIdResponse); err != nil {
		panic(err)
	}

	inventory_item_id := getIdResponse.List_InventoryItems.InventoryItems[0].Id

	currentTime := time.Now()

	updateQuantity := graphql.NewRequest(`
	mutation m(
		$id: ID!,
		$quantity: Int,
		$lastUpdated: String
	) {
		update_Inventory_async(
		  id: $id,
		  input: {
			quantity: $quantity
			lastUpdated: $lastUpdated
		  }
		) {
		  error
		  result {
			_id
		  }
		}
	  }
	`)

	updateQuantity.Var("id", inventory_item_id)
	updateQuantity.Var("quantity", quantityPtr)
	updateQuantity.Var("lastUpdated", currentTime.Format("2006-01-02T15:04:05Z"))
	updateQuantity.Header.Set("x-api-key", api_key)

	if err := client.Run(context.Background(), updateQuantity, nil); err != nil {
		panic(err)
	}

	fmt.Printf("%s quantity: %d\n", *itemNamePtr, *quantityPtr)
}

File: features/share/graphql/python3/mutation.py
from datetime import datetime
import os

import click
from gql import gql, Client
from gql.transport.requests import RequestsHTTPTransport
import urllib3

urllib3.disable_warnings()

api = os.getenv("API")
api_key = os.getenv("API_KEY")

transport = RequestsHTTPTransport(
    url=api,
    use_json=True,
    headers={
        "Content-type": "application/json",
        "x-api-key": api_key
    },
    verify=False,
    retries=3,
)

gql_client = Client(
    transport=transport,
    fetch_schema_from_transport=True,
)

def get_id(itemname):
    '''Get the _id value associated with the itemname

    Parameters
    ----------
    itemname: string, required

    Returns
    -------
    result: string
        _id value associated with the itemname
    '''

    params = {
        "itemName": itemname
    }

    query = gql(
        """
        query listInventoryItems($itemName:String) {
            list_InventoryItems(filter: {itemName: {eq: $itemName}}) {
                _InventoryItems {
                    _id
                }
            }
        }
        """
    )

    try:
        result = gql_client.execute(
            query,
            variable_values=params
        )
    except Exception as e:
        raise Exception(f'Error: {str(e)}')
    
    try:
        return(result['list_InventoryItems']['_InventoryItems'][0]['_id'])
    except:
        raise Exception(f'{itemname} not found')

def set_quantity(itemid, quantity):
    '''Update the quantity associated with the i

    Parameters
    ----------
    itemid: string, required


    Returns
    -------
    result: string
        _id value associated with the itemname
    '''

    params = {
        "id": itemid,
        "quantity": quantity,
        "lastupdated": datetime.now().strftime('%Y-%m-%dT%H:%M:%SZ')
    }

    query = gql(
        """
        mutation updateInventory($id: ID!, $quantity: Int, $lastupdated: String) {
            update_Inventory(id: $id, input: {quantity: $quantity, lastUpdated: $lastupdated}, syncMode: ASYNC) {
                transaction {
                    _id
                }
            }
        }
        """
    )

    try:
        result = gql_client.execute(
            query,
            variable_values=params
        )
    except Exception as e:
        raise Exception(f'Error: {str(e)}')
    
    try:
        return(result['update_Inventory']['result']['_id'])
    except:
        raise Exception(f'{itemid} not found')


@click.command()
@click.option('--itemname', '-i', default="Thing 1", help="Item name")
@click.option('--quantity', '-q', type=int, default=1, help="Quantity")
def update_quantity(itemname, quantity):
    item_id = get_id(itemname)

    result = set_quantity(item_id, quantity)

    if result:
        print(f'{itemname} quantity: {quantity}')


if __name__ == '__main__':
    update_quantity()
